flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java

```diff
@@ -262,4 +269,268 @@ protected ExtendedEndpoint(
+    public void testScheduleWithFixedDelay()
+            throws InterruptedException, ExecutionException, TimeoutException {
+        final RpcEndpoint endpoint = new BaseEndpoint(rpcService);
+        try {
+            endpoint.start();
+            endpoint.getMainThreadExecutor()
+                    .scheduleWithFixedDelay(
+                            endpoint::validateRunsInMainThread, 100, 100, TimeUnit.MILLISECONDS);
+            fail(
+                    "Expected to fail with a UnsupportedOperationException"
+                            + " since we have not implemented this method currently.");
+        } finally {
+            RpcUtils.terminateRpcEndpoint(endpoint, TIMEOUT);
+        }
+    }
+
+    /**
+     * test execute the callable in the main thread of the underlying RPC service, returning a
+     * future for the result of the callable. If the callable is not completed within the given
+     * timeout, then the future will be failed with a TimeoutException. This schedule method is
+     * called directly from RpcEndpoint, MainThreadExecutor do not support this method.
+     */
+    @Test
+    public void testCallAsync() throws InterruptedException, ExecutionException, TimeoutException {
+        final RpcEndpoint endpoint = new BaseEndpoint(rpcService);
+        final Time timeout = Time.milliseconds(500);
+        final Integer expectedInteger = 12345;
+        try {
+            endpoint.start();
+            CompletableFuture<Integer> integerFuture =
+                    endpoint.callAsync(
+                            () -> {
+                                endpoint.validateRunsInMainThread();
+                                return expectedInteger;
+                            },
+                            timeout);
+            assertEquals(integerFuture.get(timeout.getSize(), timeout.getUnit()), expectedInteger);
+        } finally {
+            RpcUtils.terminateRpcEndpoint(endpoint, TIMEOUT);
+        }
+    }
+
+    /**
+     * make the callable sleep some time more than specified timeout, so TimeoutException is
+     * expected.
+     */
+    @Test(expected = TimeoutException.class)
+    public void testCallAsyncTimeOut()
+            throws InterruptedException, ExecutionException, TimeoutException {
+        final RpcEndpoint endpoint = new BaseEndpoint(rpcService);
+        final Time timeout = Time.milliseconds(200);
+        final Integer expectedInteger = 12345;
+        try {
+            endpoint.start();
+            CompletableFuture<Integer> integerFuture =
+                    endpoint.callAsync(
+                            () -> {
+                                endpoint.validateRunsInMainThread();
+                                TimeUnit.MILLISECONDS.sleep(300);
+                                return expectedInteger;
+                            },
+                            timeout);
+            integerFuture.get(timeout.getSize(), timeout.getUnit());
```

Comment:
This should not work.

```
future = endpoint.callAsync(callable, timeout1);
future.get(timeout2.getSize(), timeout2.getUnit);
```
I think this case should test whether `timeout1` in the above example takes effect, while currently the `TimeoutException` comes from `timeout2`. You can verify that by replace `integerFuture.get(timeout.getSize(), timeout.getUnit())` with `integerFuture.get()`. An `ExecutionException` should be thrown.

When the callable execution takes longer than `timeout2`, `integerFuture` will be completed exceptionally. Trying to `get` a `CompletableFuture` that is completed exceptionally will result in the `ExecutionException`. Instead, we should verify that `integerFuture` is completed with a timeout exception.